File: 我的大学/algorithm2/05-动态规划/02-矩阵连乘.go
package main

import "fmt"

const maxn = 10;
var p [maxn]int
var dp [maxn][maxn]int

func min(a, b int) int {
    if a < b {
        return a
    } else {
        return b
    }
}

func main() {
    var n int
    fmt.Scanf("%d", &n)

    for i := 1; i <= n+1; i++ {
        fmt.Scanf("%d", &p[i])
    }

    for r := 2; r <= n; r++ {
        for l := r; l >= 1; l-- {
            if l == r {
                dp[l][r] = 0
            } else {
                dp[l][r] = 0x3f3f3f3f
            }
            for k := l; k < r; k++ {
                dp[l][r] = min(dp[l][r], dp[l][k] + dp[k+1][r] + p[l] * p[k+1] * p[r+1])
            }
        }
    }

    fmt.Println(dp[1][n])
}

File: 我的大学/algorithm2/05-动态规划/02-矩阵连乘.cpp
#include <cstdio>
const int maxn = 10;   // 限定最多 10 个矩阵
int p[maxn], m[maxn][maxn];

int main() {
    int n; scanf("%d", &n);

    for (int i = 0; i < n+1; ++ i)
        scanf("%d", &p[i]);   // 矩阵的维度

    for (int i = 2; i <= n; ++ i) {
        for (int j = 0; j <= n-i; ++ j) {
            int r = j + i - 1;
            m[j][r] = m[j + 1][r] + p[j] * p[j + 1] * p[r + 1];   // m[i][j]初值
            // 求最小值
            for (int k = j + 1; k < r; ++ k) {
                m[j][r] = min(m[j][r], m[j][k] + m[k + 1][r] + p[j] * p[k + 1] * p[r + 1]);
            }
        }
    }
    printf("%d\n", m[0][n - 1]);
    return 0;
}
